src/main/software/amazon/event/ruler/ACTask.java

```diff
@@ -29,11 +36,44 @@ NameState startState() {
         return machine.getStartState();
     }
 
+    ACStep nextStep() {
+        return stepQueue.remove();
+    }
+
+    /*
+     *  Add a step to the queue for later consideration
+     */
+    void addStep(final int fieldIndex, final NameState nameState, final Set<Double> candidateSubRuleIds,
+                 final ArrayMembership membershipSoFar) {
+        stepQueue.add(new ACStep(fieldIndex, nameState, candidateSubRuleIds, membershipSoFar));
+    }
+
+    boolean stepsRemain() {
+        return !stepQueue.isEmpty();
+    }
+
     List<Object> getMatchedRules() {
-        return new ArrayList<>(rules);
+        return new ArrayList<>(matchingSubRules.stream()
+                .map(r -> r.getRule())
+                .collect(Collectors.toSet()));
     }
 
-    void collectRules(final NameState nameState) {
-        rules.addAll(nameState.getRules());
+    void collectRules(final Set<Double> candidateSubRuleIds, final NameStateWithPattern nameStateWithPattern) {
+        Set<NameState.SubRule> terminalSubRules = nameStateWithPattern.getNameState().getTerminalSubRulesForPattern(
+                nameStateWithPattern.getPattern());
+        if (terminalSubRules == null) {
+            return;
+        }
+
+        // If no candidates, that means we're on the first step, so all sub-rules are candidates.
+        if (candidateSubRuleIds.isEmpty()) {
+            matchingSubRules.addAll(terminalSubRules);
+        } else {
+            for (NameState.SubRule subRule : terminalSubRules) {
```

Comment:
This is one matching path, the perf would bound to number of subrules and candidate sub rules, but I see your point, there shall have small number of sub rules in reality, shall be no visible perf diff unless we create dedicated benchmark with large quantity of sub rules. 
btw, change candidate as double set is a good enhancement

Reply:
Thanks. Yeah, this code is basically the same code as from Task. At some point, we need to reduce the duplicated code from the AC vs non-AC code paths.